ecs-agent/tmds/handlers/fault/v1/handlers/middlewares.go

```diff
@@ -0,0 +1,71 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package handlers
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+)
+
+// responseWriterWrapper is a wrapper of http.ResponseWriter.
+type responseWriterWrapper struct {
+	http.ResponseWriter
+	status int
+}
+
+func (w *responseWriterWrapper) WriteHeader(status int) {
+	w.status = status
+	w.ResponseWriter.WriteHeader(status)
+}
+
+func TelemetryMiddleware(
+	next http.Handler,
+	mf metrics.EntryFactory,
+	faultOperation, faultType string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		startTime := time.Now()
+		rw := &responseWriterWrapper{ResponseWriter: w}
+
+		// Call the next handler in the chain
+		next.ServeHTTP(rw, r)
+
+		durationInMs := time.Since(startTime).Milliseconds()
+		loggerFields := logger.Fields{
+			"StatusCode":   rw.status,
+			"DurationInMs": durationInMs,
+			"Request":      r.URL.Path,
+		}
+		logger.Info("Complete the request", loggerFields)
```

Comment:
Q: What does "Complete the request" mean here? Should we add a log here saying "Emitting metrics...."?

Reply:
It means the middleware handler has finished to process this request. How about "Finished the telemetry middleware."?

Reply:
yep sgtm, thanks!

Reply:
Nit: Can we change "Complete the request" to something that's easier to understand

Reply:
Updated.